### PR TITLE
writing tests for response and request classes 

### PR DESCRIPTION
**Pull request by ephieo**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Echo echo.Server
+# HTTP Server
```
Comment (Kotauror) [inline README.md:1]: 👍 good you pick up on possible Readme changes

File: gradlew [hunk 2]
```diff
@@ -1,7 +1,7 @@
-#!/usr/bin/env sh
+#!/bin/sh
 
 #
-# Copyright 2015 the original author or authors.
+# Copyright © 2015-2021 the original authors.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (Kotauror) [inline gradlew:6]: wow, so many changes to this file! Is this something that you write by hand or is it automatically created? What caused the changes?

File: src/main/java/http/request/Request.java [hunk 3]
```diff
@@ -1,5 +1,41 @@
 package http.request;
 
+import java.util.*;
```
Comment (Kotauror) [inline src/main/java/http/request/Request.java:3]: 👏 short and clear, this should work very well as Request's abstraction

File: src/main/java/http/request/RequestParser.java [hunk 4]
```diff
@@ -6,68 +6,10 @@
 import java.util.*;
 
 public class RequestParser {
```
Comment (Kotauror) [inline src/main/java/http/request/RequestParser.java:8]: Does the file have to stay if it has most of the content removed?

File: src/main/java/http/request/RequestRouter.java [hunk 5]
```diff
@@ -7,33 +7,41 @@
 import java.nio.charset.StandardCharsets;
 
 
-import static http.request.RequestParser.*;
-
 public class RequestRouter {
     public static void requestReader(InputStream input) throws IOException {
```
Comment (Kotauror) [inline src/main/java/http/request/RequestRouter.java:11]: hmm, I'm not sure about the naming and single responsibility principle here - the `requestReader` sounds like a function that returns an instance of the `requestReader` instead of doing an action, while the action here is reading annd parsing, I wonder if this could be split into two functions (one for reading and one for calling parse)? What do you think?

File: src/main/java/http/request/RequestRouter.java [hunk 6]
```diff
@@ -7,33 +7,41 @@
...
 
-        String response = formatRequestBody();
 
-        output.write(response.getBytes());
-        output.flush();
 
-    }
+
+
+//    public static void sendResponse(OutputStream output) throws IOException {
```
Comment (ellehallal) [inline src/main/java/http/request/RequestRouter.java]: We can delete these commented out lines :)

File: src/main/java/http/response/Response.java [hunk 7]
```diff
@@ -8,34 +8,42 @@
...
 
-    public void addBody (byte[] body){
-        setHeaders("Content-Length",Integer.toString(body.length));
-        this.body = body;
+    public Map<String, String> getHeaders() {
+        return this.headers;
     }
 
-    public void buildResponse(){
+    public void addBody(byte[] body) {
```
Comment (ellehallal) [inline src/main/java/http/response/Response.java]: Nitpick - what are your thoughts on renaming `addBody` to `setBody`, to be in line with the other methods which are setting properties (i.e `setHeaders` and `setStatus`)?

File: src/test/java/RequestTest.java [hunk 8]
```diff
@@ -35,4 +36,19 @@ public void RequestDataIsRead () throws IOException {
 
         assertEquals("http.Server received : done\ndone", output.toString().trim());
     }
+    @Test
+    public void checkThatRequestClassValuesAreAssigned (){
```
Comment (ellehallal) [inline src/test/java/RequestTest.java:30]: Good job 👍🏽

File: src/test/java/RequestTest.java [hunk 9]
```diff
@@ -16,14 +17,14 @@
...
+//    public void inputSentToClient () throws IOException {
+//        ByteArrayOutputStream mockOutput = new ByteArrayOutputStream();
+//
+//        sendResponse(mockOutput);
+//
+//        assertArrayEquals("status: 200 OK\n".getBytes(), mockOutput.toByteArray());
+//    }
 
     @Test
     public void RequestDataIsRead () throws IOException {
```
Comment (ellehallal) [inline src/test/java/RequestTest.java:20]: Is this test related to the `Request` class? If not let's move it, or delete if it's no longer needed

File: src/main/java/http/response/ResponseFormatter.java [hunk 10]
```diff
@@ -1,4 +1,29 @@
...
+    private byte[] body;
+
+    public void ResponseFormatter( int statusCode, String statusMessage, HashMap<String, String> headers, byte[] body) {
+        this.statusCode = statusCode;
+        this.statusMessage = statusMessage;
+        this.headers = headers;
+        this.body = body;
+    }
+
+    public String buildResponse (){
```
Comment (ellehallal) [inline src/main/java/http/response/ResponseFormatter.java]: Has this method been tested?

File: src/main/java/http/request/RequestRouter.java [hunk 11]
```diff
@@ -7,33 +7,41 @@
...
+    public static String[] parseRequest(String requestLine) {
+        String[] splitData = requestLine.split(" ");
+        return splitData;
     }
+ public static Request buildRequest (String method, String path, String protocol) {
+        Request request = new Request();
+        request.setMethod(method);
+        request.setPath(path);
+        request.setHeaders("protocol",protocol);
+        return request;
```
Comment (Kotauror) [inline src/main/java/http/request/RequestRouter.java:29]: Am I correct that your goal here is to:
- use the `parseRequest` and `buildRequest` to transform the inpoutStream into the `request` object (partially coded already)
- then (not coded yet) add to this class a routing function that will analyse the request object to then handle it accordingly?

File: src/main/java/http/response/Response.java [hunk 12]
```diff
@@ -8,34 +8,42 @@
...
 
-   public void setHeaders (String key, String value){
-        this.headers.put(key,value);
+    public String getStatusMessage() {
+        return this.statusMessage;
+    }
+    public int getStatusCode() {
+        return this.statusCode;
+    }
+    public void setHeaders(String key, String value) {
```
Comment (Kotauror) [inline src/main/java/http/response/Response.java]: very small nitpick, but this should be called `addHeaders` - the reason is that `set` suggests that the whole headers hashmap will be replaced (which is not a case here), while `add` that a header will be added (like you do it)

File: src/test/java/RequestRouterTest.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    public void parseRequestData() {
+        RequestRouter requestRouter = new RequestRouter();
+        String requestData = "GET /hello HTTP/1.1";
+
+        String[] splitData = requestRouter.parseRequest(requestData);
+
+        assertEquals(splitData[0],"GET" );
+        assertEquals(splitData[1],"/hello" );
+        assertEquals(splitData[2],"HTTP/1.1" );
+    }
```
Comment (Kotauror) [inline src/test/java/RequestRouterTest.java:21]: this is a great start for the request builder! This can then be used to set the properties of the request - `splitData[0]` in `setMethod`, `splitData[1]` for the path etc.

File: src/test/java/RequestTest.java (hunk 9 above)
Comment (Kotauror) [reply src/test/java/RequestTest.java:20]: I find this test useful, but it's not correctly working atm due to some changes in what you are printing and lack of new line in `mockInput` . I've changed it this way to make it work: 

```ruby
    @Test
    public void RequestDataIsRead () throws IOException {
        ByteArrayInputStream mockInput = new ByteArrayInputStream("done\n somethig To parse".getBytes());
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        System.setOut(new PrintStream(output));

        requestReader(mockInput);

        assertEquals("http.Server received : done\noutput read", output.toString().trim());
    }
```

File: gradlew (hunk 2 above)
Comment (ephieo) [reply gradlew:6]: No this file is auto-generated by gradle

File: src/main/java/http/request/Request.java (hunk 3 above)
Comment (ephieo) [reply src/main/java/http/request/Request.java:3]: Thank you 👍

File: src/main/java/http/request/RequestParser.java (hunk 4 above)
Comment (ephieo) [reply src/main/java/http/request/RequestParser.java:8]: I'm not sure if i'll use it or not I think it fits in just not sure where yet.

File: src/main/java/http/request/RequestRouter.java (hunk 6 above)
Comment (ephieo) [reply src/main/java/http/request/RequestRouter.java]: 👍

File: src/main/java/http/response/Response.java (hunk 7 above)
Comment (ephieo) [reply src/main/java/http/response/Response.java]: Yes, I don't mind I'll make that change.

File: src/test/java/RequestTest.java (hunk 8 above)
Comment (ephieo) [reply src/test/java/RequestTest.java:30]: thank you 👍

File: src/test/java/RequestTest.java (hunk 9 above)
Comment (ephieo) [reply src/test/java/RequestTest.java:20]: I had to change it slightly but it works now thank you 😄

File: src/main/java/http/response/ResponseFormatter.java (hunk 10 above)
Comment (ephieo) [reply src/main/java/http/response/ResponseFormatter.java]: it's old code that needs to be deleted, i think i deleted the test files fo it.

File: src/main/java/http/request/RequestRouter.java (hunk 11 above)
Comment (ephieo) [reply src/main/java/http/request/RequestRouter.java:29]: yes that's my plan.

File: src/main/java/http/response/Response.java (hunk 12 above)
Comment (ephieo) [reply src/main/java/http/response/Response.java]: 👍

File: src/test/java/RequestRouterTest.java (hunk 13 above)
Comment (ephieo) [reply src/test/java/RequestRouterTest.java:21]: thank you 👍
